components/base/kinematicbase/execution.go

```diff
@@ -0,0 +1,471 @@
+			// nonzero time to run especially when using the localizer.
+			actualTimeElapsed := time.Since(arcStartTime)
+			remainingTimeStep := time.Duration(1000*1000*timeElapsed)*time.Microsecond - actualTimeElapsed
+
+			if remainingTimeStep > 0 {
+				utils.SelectContextOrWait(ctx, remainingTimeStep)
+				if ctx.Err() != nil {
+					return tryStop(ctx.Err())
+				}
+			}
+			distIncVel := step.linVelMMps.Y
+			if distIncVel == 0 {
+				distIncVel = step.angVelDegps.Z
+			}
+			currentInputs := []referenceframe.Input{
+				step.arcSegment.StartConfiguration[0],
+				step.arcSegment.StartConfiguration[1],
+				{step.arcSegment.StartConfiguration[2].Value + math.Abs(distIncVel)*timeElapsed},
+			}
+			ptgk.inputLock.Lock()
+			ptgk.currentInputs = currentInputs
+			ptgk.inputLock.Unlock()
+
+			// If we have a localizer, we are able to attempt to correct to stay on the path.
+			// For now we do not try to correct while in a correction.
+			if ptgk.Localizer != nil {
+				actualPose, err := ptgk.Localizer.CurrentPosition(ctx)
+				if err != nil {
+					return err
+				}
+				expectedPoseRel, err := ptgk.frame.Transform(currentInputs)
+				if err != nil {
+					return err
+				}
+
+				// This is where we expected to be on the trajectory
+				expectedPose := spatialmath.Compose(step.arcSegment.StartPosition, expectedPoseRel)
+
+				// This is where actually are on the trajectory
+				poseDiff := spatialmath.PoseBetween(actualPose.Pose(), expectedPose)
+
+				allowableDiff := ptgk.linVelocityMMPerSecond * inputUpdateStepSeconds * (minDeviationToCorrectPct / 100)
+				ptgk.logger.Debug("allowable diff ", allowableDiff, " diff now ", poseDiff.Point().Norm())
+				if poseDiff.Point().Norm() > allowableDiff || poseDiff.Orientation().AxisAngles().Theta > 0.25 {
+					ptgk.logger.Debug("expected to be at ", spatialmath.PoseToProtobuf(expectedPose))
+					ptgk.logger.Debug("Localizer says at ", spatialmath.PoseToProtobuf(actualPose.Pose()))
+					// Accumulate list of points along the path to try to connect to
+					goals := ptgk.makeCourseCorrectionGoals(
+						goalsToAttempt,
+						actualPose.Pose(),
+					)
+
+					ptgk.logger.Debug("wanted to attempt ", goalsToAttempt, " goals, got ", len(goals))
+					// Attempt to solve from `actualPose` to each of those points
+					solution, err := ptgk.courseCorrect(ctx, goals)
+					if err != nil {
+						ptgk.logger.Debug(err)
+					}
+					if solution.Solution != nil {
+						ptgk.logger.Debug("successful course correction", solution.Solution)
+
+						correctiveArcSteps := []arcStep{}
+						for i := 0; i < len(solution.Solution); i += 2 {
+							// We've got a course correction solution. Swap out the relevant arcsteps.
+							correctiveTraj, err := ptgk.ptgs[ptgk.courseCorrectionIdx].Trajectory(
+								solution.Solution[i].Value,
+								solution.Solution[i+1].Value,
+								stepDistResolution,
+							)
+							if err != nil {
+								ptgk.logger.Warn(err)
```

Comment:
[q]: why is ok to ignore error here, wouldn't we be dealing with non-sense `trajNodes`?